### ISSUE TITLE
installation of package ‘igraph’ had non-zero exit status

### ISSUE THREAD

**Issue by liuyifang** [igraph/rigraph#213]
Hi
The igraph r package has the same bug as igraph in https://github.com/igraph/igraph/commit/0ddc7727381d69eb54a17fa27114554712f6f5d5

> foreign-graphml.c: In function ‘igraph_write_graph_graphml’:
foreign-graphml.c:1408:46: error: expected ‘)’ before ‘GRAPHML_NAMESPACE_URI’
   ret=fprintf(outstream, "<graphml xmlns=\"" GRAPHML_NAMESPACE_URI "\"\n");
                                              ^
foreign-graphml.c:1412:59: error: expected ‘)’ before ‘GRAPHML_NAMESPACE_URI’
   ret=fprintf(outstream, "         xsi:schemaLocation=\"" GRAPHML_NAMESPACE_URI "\n");
                                                           ^
foreign-graphml.c:1414:38: error: expected ‘)’ before ‘GRAPHML_NAMESPACE_URI’
   ret=fprintf(outstream, "         " GRAPHML_NAMESPACE_URI "/1.0/graphml.xsd\">\n");
                                      ^
make: *** [foreign-graphml.o] Error 1
ERROR: compilation failed for package ‘igraph’

**Comment by gaborcsardi** [comment]
This is unfortunately a bug in igraph. To work around it you need a newer compiler, or wait for the next igraph release.

**Comment by liuyifang** [comment]
Thanks for your information. 
Could you tell the next igraph release date, if it will be soon, I can wait.

**Comment by gaborcsardi** [comment]
I would do it today if I could, but CRAN does not like that. But surely within two weeks I would say.

Btw. I think you can also apply this patch: https://github.com/igraph/igraph/pull/1020/files if you don't want to wait.

**Comment by liuyifang** [comment]
Thanks for your advice, I manually update the source code and solved this bug. But I meet another issue:

> Error: package or namespace load failed for ‘igraph’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/home/liuyf/R/R_LIBS/igraph/libs/igraph.so':
  libiconv.so.2: cannot open shared object file: No such file or directory
Error: loading failed
Execution halted
ERROR: loading failed

My R Environment:

> 

> CentOS Linux release 7.3.1611 (Core) 
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-11)
R 3.4.1

**Comment by liuyifang** [comment]
Problem solved.
This is because my CentOS system does not contain libiconv.so.2 shared library.
To solve this:

> wget https://ftp.gnu.org/pub/gnu/libiconv/libiconv-1.15.tar.gz
> tar -zxvf libiconv-1.15.tar.gz
> cd libiconv-1.15
> ./configure --prefix=/usr/local
> make
> make install
> 

Then

> export LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH

**Comment by arpitghiya** [comment]
Hello

I have followed your suggestion and applied the patch in the source code. Although, this time I am getting a different error while installing igraph. Can you please help.

src/foreign-gml-lexer.l:102: warning: ignoring return value of ‘fwrite’, declared with attribute warn_unused_result
gcc -m64 -std=gnu99 -I/usr/include/R -DNDEBUG   -I/usr/local/include  -DUSING_R -I. -Iinclude -Ics -Iglpk -Iplfit -ICHOLMOD/Include -IAMD/Include -ICOLAMD/Include -ISuiteSparse_config -I/usr/local/include -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -fpic -fPIC -DNDEBUG -DNPARTITION -DNTIMER -DNCAMD -DNPRINT -DPACKAGE_VERSION=\"1.1.1\" -DINTERNAL_ARPACK -DIGRAPH_THREAD_LOCAL=/**/ -fpic  -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -fpic -fPIC   -c foreign-gml-parser.c -o foreign-gml-parser.o
In file included from src/foreign-gml-parser.y:59:
igraph_hacks_internal.h:42:1: warning: "strdup" redefined
In file included from /usr/include/string.h:637,
                 from src/foreign-gml-parser.y:54:
/usr/include/bits/string2.h:1316:1: warning: this is the location of the previous definition
gcc -m64 -std=gnu99 -I/usr/include/R -DNDEBUG   -I/usr/local/include  -DUSING_R -I. -Iinclude -Ics -Iglpk -Iplfit -ICHOLMOD/Include -IAMD/Include -ICOLAMD/Include -ISuiteSparse_config -I/usr/local/include -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -fpic -fPIC -DNDEBUG -DNPARTITION -DNTIMER -DNCAMD -DNPRINT -DPACKAGE_VERSION=\"1.1.1\" -DINTERNAL_ARPACK -DIGRAPH_THREAD_LOCAL=/**/ -fpic  -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -fpic -fPIC   -c foreign-graphml.c -o foreign-graphml.o
foreign-graphml.c: In function ‘igraph_write_graph_graphml’:
foreign-graphml.c:1408: error: ‘GRAPHML_NAMESPACE_URI’ undeclared (first use in this function)
foreign-graphml.c:1408: error: (Each undeclared identifier is reported only once
foreign-graphml.c:1408: error: for each function it appears in.)
make: *** [foreign-graphml.o] Error 1
ERROR: compilation failed for package ‘igraph’
* removing ‘/usr/lib64/R/library/igraph’

**Comment by liuyifang** [comment]
Hi
You need also update `src/foreign-graphml.c` following this link: https://github.com/igraph/igraph/commit/9acfa54fa6b3d182fe458434a497f4e9b5c39955
Then it will be ok.

**Comment by arpitghiya** [comment]
Thanks! This helped

**Comment by SKrPl** [comment]
@gaborcsardi I am still facing the same problem while installing.

**Comment by Norali81** [comment]
Hi, is there anywhere I can find instructions on how to apply the patch? Thanks a lot!

**Comment by liuyifang** [comment]
You can just change the source code `src/foreign-graphml.c` manually following https://github.com/igraph/igraph/pull/1020/files and https://github.com/igraph/igraph/commit/9acfa54fa6b3d182fe458434a497f4e9b5c39955
then compile igraph following this http://kbroman.org/pkg_primer/pages/build.html

**Comment by Norali81** [comment]
Great! Thanks so much!

**Comment by colinwxl** [comment]
My system is Ubutu16.04LTS, I've solved this problem by `sudo apt-get install libxml2-dev`, refer to http://www.genek.tv/article/40

**Comment by sylv704** [comment]
On RHEL, I had to install libxml2-devel (and maybe libssh2-devel too) for the package to compile and get installed.

**Comment by whitleyo** [comment]
Hi,

I get a similar problem when installing igraph in R: 

Error: package or namespace load failed for ‘igraph’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/home/owenwhitley/R/x86_64-pc-linux-gnu-library/3.4/igraph/libs/igraph.so':
  libgfortran.so.4: cannot open shared object file: No such file or directory
Error: loading failed
Execution halted

Looking through the console, I get these warnings:

In file included from src/foreign-gml-parser.y:54:0:
igraph_hacks_internal.h:42:0: warning: "strdup" redefined [enabled by default]
 #  define strdup igraph_i_strdup
 ^
In file included from /usr/include/string.h:635:0,
                 from src/foreign-gml-parser.y:49:
/usr/include/x86_64-linux-gnu/bits/string2.h:1291:0: note: this is the location of the previous definition
 #   define strdup(s) __strdup (s)

In file included from src/foreign-lgl-parser.y:49:0:
igraph_hacks_internal.h:42:0: warning: "strdup" redefined [enabled by default]
 #  define strdup igraph_i_strdup
 ^
In file included from /usr/include/string.h:635:0,
                 from src/foreign-lgl-parser.y:48:
/usr/include/x86_64-linux-gnu/bits/string2.h:1291:0: note: this is the location of the previous definition
 #   define strdup(s) __strdup (s)
 ^

In file included from src/foreign-ncol-parser.y:49:0:
igraph_hacks_internal.h:42:0: warning: "strdup" redefined [enabled by default]
 #  define strdup igraph_i_strdup
 ^
In file included from /usr/include/string.h:635:0,
                 from src/foreign-ncol-parser.y:48:
/usr/include/x86_64-linux-gnu/bits/string2.h:1291:0: note: this is the location of the previous definition
 #   define strdup(s) __strdup (s)
 ^

In file included from hacks.c:27:0:
igraph_hacks_internal.h:42:0: warning: "strdup" redefined [enabled by default]
 #  define strdup igraph_i_strdup
 ^
In file included from /usr/include/string.h:635:0,
                 from /usr/include/memory.h:29,
                 from hacks.c:24:
/usr/include/x86_64-linux-gnu/bits/string2.h:1291:0: note: this is the location of the previous definition
 #   define strdup(s) __strdup (s)
 ^

Here's the session info:

sessionInfo()
R version 3.4.4 (2018-03-15)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.5 LTS

Matrix products: default
BLAS: /usr/lib/libblas/libblas.so.3.0
LAPACK: /usr/lib/lapack/liblapack.so.3.0

locale:
 [1] LC_CTYPE=en_CA.UTF-8       LC_NUMERIC=C               LC_TIME=en_CA.UTF-8        LC_COLLATE=en_CA.UTF-8    
 [5] LC_MONETARY=en_CA.UTF-8    LC_MESSAGES=en_CA.UTF-8    LC_PAPER=en_CA.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_CA.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.4.4 tools_3.4.4    yaml_2.1.18  

Would it be recommended that I upgrade to Ubuntu 16.04?

Thanks!

**Comment by liuyifang** [comment]
Hi, I recommend you try install libgfortran following this link: https://stackoverflow.com/questions/46516394/how-to-install-libgfortran-so-4-on-ubuntu-16-06

**Comment by EricPeter** [comment]
Install those packages
sudo apt-get install libpoppler-cpp-dev   
sudo apt-get install libapparmor-dev

**Comment by bplatt21** [comment]
Hey all, super new to this, so go easy! I am getting this error, which I have no idea what it is. 

`make: gfortran-4.8: No such file or directory
make: *** [AMD/Source/amd.o] Error 1
ERROR: compilation failed for package ‘igraph’
* removing ‘/Library/Frameworks/R.framework/Versions/3.3/Resources/library/igraph’
* restoring previous ‘/Library/Frameworks/R.framework/Versions/3.3/Resources/library/igraph’

The downloaded source packages are in
	‘/private/var/folders/x8/kpthc3ks2tg5g85hgf30mnkc0000gp/T/RtmpTtA3oc/downloaded_packages’
Warning message:
In install.packages("igraph") :
  installation of package ‘igraph’ had non-zero exit status`

I've tried all the suggestions above to solve the `non-zero exit status` issue but nothing is working. I also tried to download the package suggested in the error. /private/var/folders/x8/kpthc3ks2tg5g85hgf30mnkc0000gp/T/RtmpTtA3oc/downloaded_packages

Also, I tried to do some of the `sudo apt-` commands, and I get

`Error: unexpected symbol in "sudo apt"`

Nothing is working, what am I doing wrong?

**Comment by gaborcsardi** [comment]
You need a fortran compiler.

**Comment by gaborcsardi** [comment]
> Ok thanks, any recommendations on which one?

@bplatt21 I don't know, it depends on your system. Btw. commenting on _closed_  issues is not the best way to reach the developers.... open a  new issue  if you have problems.

**Comment by D-Harouni** [comment]
Followed some suggestion but still getting the following: 

Error: package or namespace load failed for ‘igraph’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/home/elharoun/R/x86_64-pc-linux-gnu-library/4.0/00LOCK-igraph/00new/igraph/libs/igraph.so':
  libgfortran.so.4: cannot open shared object file: No such file or directory
Error: loading failed
Execution halted

**Comment by shubham1637** [comment]
Use the following commands to install libgfortran
```
sudo add-apt-repository ppa:jonathonf/gcc-7.1
sudo apt-get update
sudo apt-get install gcc-7 g++-7
sudo apt-get install gfortran-7
```

**Comment by yuxiaokang-source** [comment]
I get the same error( my environment is ununtu20.04,R 4.1.0), I run 
`sudo apt install gfortran-7`  in terminal, then I run  `install.packages("igraph")` in rstudio, it run successfully

**Comment by taosun112** [comment]
> I get the same error( my environment is ununtu20.04,R 4.1.0), I run `sudo apt install gfortran-7` in terminal, then I run `install.packages("igraph")` in rstudio, it run successfully
我也出现了同样的问题，可以帮助一下我吗？
stao1211@163.com
